volatility3/framework/symbols/linux/utilities/modules.py

```diff
@@ -346,52 +268,30 @@ def run_modules_scanners(
         Returns:
             Dictionary mapping each plugin to its corresponding result
         """
-
-        module_gatherers = {
-            Modules.source_kernel_identifier: cls.get_kernel_module_info,
-            Modules.source_lsmod_identifier: cls._get_list_modules,
-            Modules.source_sysfs_identifier: cls._get_sysfs_modules,
-        }
+        # Throws ValueError if invalid gatherers sent in
+        Modules._validate_gatherers(caller_wanted_gatherers)
 
         kernel = context.modules[kernel_module_name]
 
         address_mask = context.layers[kernel.layer_name].address_mask
 
-        wanted_sources = Modules._validated_sources(caller_wanted_sources)
-
-        run_results = {}
-
-        run_hidden_modules = False
-
-        # Special case hidden modules since it gathers modules on its own
-        if Modules.source_hidden_identifier in wanted_sources:
-            run_hidden_modules = True
-            wanted_sources.remove(Modules.source_hidden_identifier)
-
-        # Walk each source, gathering modules
-        for wanted_source in wanted_sources:
-            run_results[wanted_source] = []
+        run_results: Dict[ModuleGathererInterface, List[ModuleInfo]] = {}
 
-            gatherer = module_gatherers[wanted_source]
+        # Walk each source gathering modules
+        for gatherer in caller_wanted_gatherers:
+            run_results[gatherer] = []
 
             # process each module coming from back the current source
-            for module in gatherer(context, kernel_module_name):
+            for module in gatherer.gather_modules(context, kernel_module_name):
+
                 # the kernel sends back a ModuleInfo directly
-                if wanted_source == Modules.source_kernel_identifier:
+                if gatherer == ModuleGathererKernel:
```

Comment:
Again, we were trying to avoid the special casing?  Why do the `ModuleGathererKernel` results need handling differently?  If we're defining this method, can't we just get it to play ball like the rest of the gatherers?

Reply:
Ok, so... would it not be better to have the interface require that a `modinfo` is returned, rather than a module-for-all-but-one-special-case?  Making specific unique things inside the generic handler breaks (because someone else might come up with a situation where they need to special case stuff, and they'll run into the same issue).  Pushing the class call into each of the yields seems a better option...

Reply:
Then every gatherer would have do the `get_module_info_for_module` loop instead of just the generic gathering knowing the kernel doesn't have a real module.. it would make really ugly code just to avoid a check on the kernel.

Reply:
I don't think that's what I was suggesting?  Just that instead of `return thing`, they each now do `return cls.get_module_info_for_module(address_mask, thing)`?  I guess it might require feeding the `address_mask` in though...

Reply:
They all are generators, so each would have to make that call to `cls.get_module_info_for_module` for each element. Check for example how lsmod looks now here, it would need to loop the calls (which would be copy/paste in every gatherer):

https://github.com/volatilityfoundation/volatility3/blob/a0f3cba6f6d71648a4feb884502491b773815255/volatility3/framework/symbols/linux/utilities/modules.py#L547

Reply:
The other option is *not* to special case this, by the module, but based on the type returned in the yield?  That was you handle a gatherer that decides to just return a modinfo (as allowed by the interface) and nothing's been special cased...

Reply:
So an `isinstance(thing, Module) then`...